hop/publish.py

```diff
@@ -4,64 +4,38 @@
 __description__ = "tools to parse and publish messages"
 
 
-import argparse
-import warnings
-
+from .auth import load_auth
 from . import cli
-from .io import Stream
-from .models import GCNCircular, VOEvent, MessageBlob
+from . import io
 
 
 # ------------------------------------------------
 # -- main
 
 
 def _add_parser_args(parser):
-    cli.add_url_opts(parser)
+    cli.add_client_opts(parser)
     parser.add_argument(
         "message", metavar="MESSAGE", nargs="+", help="One or more messages to publish.",
     )
-    cli.add_config_opts(parser)
-
     parser.add_argument(
         "-f",
         "--format",
-        type=str,
-        default="blob",
-        help="Specifies the format of the message(s), such as 'circular' or 'voevent'. "
-        "Specify 'blob' if sending an unstructured message. Default: 'blob'.",
+        choices=io.Deserializer.__members__,
+        default=str(io.Deserializer.BLOB),
+        help="Specify the message format. Defaults to BLOB for an unstructured message.",
     )
 
 
-def _main(args=None):
+def _main(args):
     """Parse and publish messages.
 
     """
+    auth = load_auth() if not args.no_auth else None
+    loader = io.Deserializer[args.format]
+    stream = io.Stream(auth=auth)
 
-    if not args:
-        parser = argparse.ArgumentParser()
-        _add_parser_args(parser)
-        args = parser.parse_args()
-
-    # load config if specified
-    config = cli.load_config(args)
-
-    # create map of possible message formats and their model loader
-    model_loader = {
-        "circular": GCNCircular.from_email_file,
-        "voevent": VOEvent.from_xml_file,
-        "blob": MessageBlob.from_text,
-    }
-
-    stream = Stream(format="json", config=config)
     with stream.open(args.url, "w") as s:
-        if args.format in model_loader:
-            loader = model_loader[args.format]
-        else:
-            warnings.warn(
-                "Warning: format not recognized. Sending as unstructured blob")
-            loader = model_loader["blob"]
-
         for message_file in args.message:
-            message_model = loader(message_file)
-            s.write(message_model.wrap_message())
+            message = loader.load_file(message_file)
```

Comment:
This forces the published message to be a file -- is this desired? This precludes us from sending string messages from CLI:
```
hop publish kafka://scimma-server:9092/test-blob-string01 -f BLOB "plain text string"
Traceback (most recent call last):
  File "/opt/conda/bin/hop", line 10, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.7/site-packages/hop/__main__.py", line 72, in main
    args.func(args)
  File "/opt/conda/lib/python3.7/site-packages/hop/publish.py", line 40, in _main
    message = loader.load_file(message_file)
  File "/opt/conda/lib/python3.7/site-packages/hop/io.py", line 145, in load_file
    return self.value.load_file(input_file)
  File "/opt/conda/lib/python3.7/site-packages/hop/models.py", line 226, in load_file
    with open(filename, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'plain text string'
```

We may not want this anyways, but checking just in case (especially since `models` has the tooling to read strings via `load()`).

Reply:
That's a good point, I hadn't appreciated that's how you wired it up when you added the blob functionality. I think being able to pass in strings is good to have, but I'm a bit wary of having this functionality present with one format but not another without it being more clear how it's being handled. As in, I think it may be clearer if `load()` has consistent behavior across all messages.

What may be nice is to be able to pipe input into `hop publish` in the same way other UNIX commands are done and this may tie nicely into passing in strings, as in:
`cat stringlist.txt | hop publish kafka://host:port/topic`

This would in principle work with all formats, as long as each string token is a message to be published. I think what we'd need to do is to modify the message argparse option to possibly take in zero arguments as well and check if `sys.stdin.isatty()` to read terminal input from stdin. What do you think?

Reply:
I agree, and I think implementing the UNIX-like piping method would be nice. I think it is out-of-scope for this PR, so I made #86.